### DISCUSSION TITLE
data access after user logout

### DISCUSSION THREAD

**Discussion by kaioker** [ory/hydra#3388] [Q&A]
we need a way of accessing a user's scopes after the user has left the system, assuming they have consented to the use of their info while offline at the consent screen.

example usage: _user A_ signs into a client website using oauth, granting scopes from the user database, and leaves a comment on _user B_'s post.
user _A then_ logs out. when _user B_ opens their page, they should see _user A_'s name and profile photo next to their comment.

as the user information is stored on the user database system, the client service needs a way to display _user A_'s information to _user B_ without _user A_ being online.

overall there are three systems in play;
1. the client services, i.e. the web applications
2. the ory hydra + openid connect server
3. the user database and authentication system

the user database contains information about the user such as name, email, and profile photo. each of these are available as scopes that can be granted to the multiple client services based on the consent provided at login through the oauth2 server

**attempted solutions:**

first we attempted to store the access and refresh tokens in a database, however if the user refreshed the token themselves the server stored access was invalidated, and if the server refreshed the token, the user was invalidated. users signing in from multiple devices got disconnected from all devices and the server.

second we tried to run the login flow twice. once for the user, and follow it with a second for the client. this resulted in lockouts as the second request nulls the first request's tokens, causing an infinite redirect loop.

third we tried logging in the user, storing the credentials, then abandoning the session and logging in again with a different session id. the ory hydra server detected this as replay abuse and locked all the tokens causing the same issues as before.

**at this point we see only two remaining options:**

1. each individual client service caches the user's information and refreshes the cache every user login. - undesirable as this is a lot of duplicate data everywhere, all of it slowly drifting out of sync over time
2. the user database has an api to retrieve user information. - undesirable as this feels like it defeats the point of oauth in the first place and also ignores the user consents granted through the oauth server

**preferred solution:**

a way for the client to get the user information from the database without interrupting the access and refresh tokens the user is using to remain logged in. this needs to be done securely and respect the user's granted scopes. ideally this is also disconnected when the user rejects all outstanding tokens.

for example: _user B_ has decided they no longer want to be associated with a particular client service and so rejects all active tokens for that client through the oauth server. this should sever the connection between the client service and the user information database, but leaves the connection to other client services that still have consent alive

are we going about this the wrong way? how do services like google and facebook do this? i have a google account and my profile picture is available on comments on youtube or wherever i have signed in using it, and it updates everywhere within a day if i change my profile pic or display name so it must be possible in some capacity.

**Comment by vinckr** [comment]
Hello @kaioker 
Apologies for the very late answer!

It seems like you're trying to implement a system where user information can be accessed even when the user is offline, and this information should be updated across all client services when changes are made.
In OAuth2 and OpenID Connect, the concept of scopes is used to specify the level of access that a client has to a user's resources. When a user logs in and grants consent, they can specify the scopes (or permissions) that the client has. These scopes can include access to user information such as name, email, and profile photo.

Tthe challenge here is to maintain this access even when the user is offline, and to ensure that any changes to the user information are reflected across all client services.

One possible solution could be to use refresh tokens. In OAuth2 and OpenID Connect, when an access token expires, a refresh token can be used to obtain a new access token without the user's involvement. This could potentially allow you to maintain access to the user's information even when they are offline.

However, it's important to note that the use of refresh tokens needs to be done securely and in a way that respects the user's granted scopes. Also, when a user revokes consent, the connection to the user information should be severed.

As for how services like Google and Facebook do this, they likely have complex custom systems in place to manage user information and consent across multiple services. They may also use APIs to retrieve user information, although this would need to be done in a way that respects user consent and the principles of OAuth.

Unfortunately, I dont know a solution to your specific problem. I would recommend looking into the official OAuth2 and OpenID Connect specifications, as well as the Ory Hydra documentation and possibly source code for more information on how to implement this.

My Sources:
[OAuth2 fundamentals](https://www.ory.sh/docs/oauth2-oidc/overview/oauth2-concepts)
[OAuth2 refresh token grant](https://www.ory.sh/docs/oauth2-oidc/refresh-token-grant)
[Ory Hydra case study](https://www.ory.sh/docs/hydra/case-study)
[Custom login and consent flow](https://www.ory.sh/docs/oauth2-oidc/custom-login-consent/flow)
